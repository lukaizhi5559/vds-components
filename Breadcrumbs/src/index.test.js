import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { Breadcrumbs, BreadcrumbItem } from '.';
import { ColorTokens } from '@vds-tokens/color';

function rgbToHex(rgb) {
  rgb = rgb
    .substr(4)
    .split(')')[0]
    .split(', ');
  let r = (+rgb[0]).toString(16),
    g = (+rgb[1]).toString(16),
    b = (+rgb[2]).toString(16);
  if (r.length == 1) r = '0' + r;
  if (g.length == 1) g = '0' + g;
  if (b.length == 1) b = '0' + b;
  return '#' + r + g + b;
}

test('<Breadcrumbs>', () => {
  const { getByText, container } = render(
    <Breadcrumbs>
      <BreadcrumbItem ariaLabel="Go to Crumb 1" href="#">
        First default crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 2" href="#" disabled>
        Disabled crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 3" href="#">
        Second default crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 4" href="#">
        Selected crumb
      </BreadcrumbItem>
    </Breadcrumbs>
  );

  // Run tests on the fireEvents on BreadcrumbAnchor
  const breadcrumbAnchor = container.querySelector('a');
  fireEvent.click(breadcrumbAnchor);
  fireEvent.mouseDown(breadcrumbAnchor);

  // get rendered Breadcrumb elements to test against
  const firstDefaultBreadcrumb = getByText('First default crumb');
  const secondDefaultBreadcrumb = getByText('Second default crumb');
  const disabledBreadcrumb = getByText('Disabled crumb');
  const selectedBreadcrumb = getByText('Selected crumb');

  // Breadcrumb elements that are "disabled" and the last "selected" Breadcrumb should be <span> elements
  expect(disabledBreadcrumb.tagName.toLowerCase()).toBe('span');
  const disabledBreadcrumbStyles = window.getComputedStyle(disabledBreadcrumb)
    ._values;
  expect(rgbToHex(disabledBreadcrumbStyles.color)).toBe(
    ColorTokens.elements.secondary.onlight.value
  );
  // selected breadcrumb style
  expect(selectedBreadcrumb.tagName.toLowerCase()).toBe('span');
  const selectedBreadcrumbStyles = window.getComputedStyle(selectedBreadcrumb)
    ._values;
  expect(rgbToHex(selectedBreadcrumbStyles.color)).toBe(
    ColorTokens.elements.primary.onlight.value
  );

  expect(selectedBreadcrumbStyles['font-weight']).toBe('700');

  // default breadcrumb styles
  expect(firstDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
  const firstDefaultBreadcrumbStyles = window.getComputedStyle(
    firstDefaultBreadcrumb
  )._values;
  expect(firstDefaultBreadcrumbStyles['font-family']).toBe('Verizon-NHG-eTX');
  expect(firstDefaultBreadcrumbStyles['font-size']).toBe('0.75rem');
  expect(firstDefaultBreadcrumbStyles['font-weight']).toBe('400');
  expect(firstDefaultBreadcrumbStyles['line-height']).toBe('1rem');
  expect(firstDefaultBreadcrumb).toHaveStyleRule(
    'color',
    ColorTokens.elements.primary.onlight.value
  );

  expect(secondDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
  expect(container.firstChild).toMatchSnapshot();
});

test('<Breadcrumbs inverted bg>', () => {
  const { getByText, container } = render(
    <Breadcrumbs surface="dark">
      <BreadcrumbItem ariaLabel="Go to Crumb 1" href="#">
        First default crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 2" href="#" disabled>
        Disabled crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 3" href="#">
        Second default crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 4" href="#">
        Selected crumb
      </BreadcrumbItem>
    </Breadcrumbs>
  );

  // Run tests on the fireEvents on BreadcrumbAnchor
  const breadcrumbAnchor = container.querySelector('a');
  fireEvent.click(breadcrumbAnchor);
  fireEvent.mouseDown(breadcrumbAnchor);

  // get rendered Breadcrumb elements to test against
  const firstDefaultBreadcrumb = getByText('First default crumb');
  const secondDefaultBreadcrumb = getByText('Second default crumb');
  const disabledBreadcrumb = getByText('Disabled crumb');
  const selectedBreadcrumb = getByText('Selected crumb');

  // Breadcrumb elements that are "disabled" and the last "selected" Breadcrumb should be <span> elements
  expect(disabledBreadcrumb.tagName.toLowerCase()).toBe('span');
  const disabledBreadcrumbStyles = window.getComputedStyle(disabledBreadcrumb)
    ._values;
  expect(disabledBreadcrumb).toHaveStyleRule('color', '#a7a7a7');

  // selected breadcrumb style
  expect(selectedBreadcrumb.tagName.toLowerCase()).toBe('span');
  const selectedBreadcrumbStyles = window.getComputedStyle(selectedBreadcrumb)
    ._values;
  expect(selectedBreadcrumb).toHaveStyleRule('color', '#ffffff');
  expect(selectedBreadcrumbStyles['font-weight']).toBe('700');

  // default breadcrumb styles
  expect(firstDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
  const firstDefaultBreadcrumbStyles = window.getComputedStyle(
    firstDefaultBreadcrumb
  )._values;
  expect(firstDefaultBreadcrumb).toHaveStyleRule('color', '#ffffff');
  expect(firstDefaultBreadcrumbStyles['font-family']).toBe('Verizon-NHG-eTX');
  expect(firstDefaultBreadcrumbStyles['font-size']).toBe('0.75rem');
  expect(firstDefaultBreadcrumbStyles['font-weight']).toBe('400');
  expect(firstDefaultBreadcrumbStyles['line-height']).toBe('1rem');
  expect(secondDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
});

test('<VDS Breadcrumbs>', () => {
  const { getByText, container } = render(
    <Breadcrumbs surface="dark">
      <BreadcrumbItem ariaLabel="Go to Crumb 1" href="#">
        First default crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 2" href="#" disabled>
        Disabled crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 3" href="#">
        Second default crumb
      </BreadcrumbItem>
      <BreadcrumbItem href="#">Selected crumb</BreadcrumbItem>
    </Breadcrumbs>
  );

  // Run tests on the fireEvents on BreadcrumbAnchor
  const breadcrumbAnchor = container.querySelector('a');

  fireEvent.click(breadcrumbAnchor);
  fireEvent.mouseDown(breadcrumbAnchor);

  // get rendered Breadcrumb elements to test against
  const firstDefaultBreadcrumb = getByText('First default crumb');
  const secondDefaultBreadcrumb = getByText('Second default crumb');
  const disabledBreadcrumb = getByText('Disabled crumb');
  const selectedBreadcrumb = getByText('Selected crumb');

  // Breadcrumb elements that are "disabled" and the last "selected" Breadcrumb should be <span> elements
  expect(disabledBreadcrumb.tagName.toLowerCase()).toBe('span');
  const disabledBreadcrumbStyles = window.getComputedStyle(disabledBreadcrumb)
    ._values;
  expect(rgbToHex(disabledBreadcrumbStyles.color)).toBe(
    ColorTokens.elements.secondary.ondark.value
  );

  const firstDefaultBreadcrumbStyless = window.getComputedStyle(
    firstDefaultBreadcrumb
  )._values;

  expect(firstDefaultBreadcrumb).toHaveStyleRule('color', '#ffffff');
  expect(firstDefaultBreadcrumbStyless['font-family']).toBe('Verizon-NHG-eTX');
  expect(firstDefaultBreadcrumbStyless['font-size']).toBe('0.75rem');
  expect(firstDefaultBreadcrumbStyless['font-weight']).toBe('400');
  expect(firstDefaultBreadcrumbStyless['line-height']).toBe('1rem');
  expect(selectedBreadcrumb.tagName.toLowerCase()).toBe('span');
  const selectedBreadcrumbStyles = window.getComputedStyle(selectedBreadcrumb)
    ._values;
  expect(selectedBreadcrumb).toHaveStyleRule(
    'color',
    ColorTokens.elements.primary.ondark.value
  );
  expect(selectedBreadcrumbStyles['font-weight']).toBe('700');

  // Breadcrumb elements that have not been "disabled" and the last Breadcrumb should be <a> elements
  expect(firstDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
  expect(firstDefaultBreadcrumb).toHaveStyleRule(
    'color',
    ColorTokens.elements.primary.ondark.value
  );

  expect(secondDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
});

test('<VDS Breadcrumbs last breadcrumb not selected>', () => {
  const { getByText, container } = render(
    <Breadcrumbs>
      <BreadcrumbItem ariaLabel="Go to Crumb 1" href="#">
        First default crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 2" href="#" disabled>
        Disabled crumb
      </BreadcrumbItem>
      <BreadcrumbItem ariaLabel="Go to Crumb 3" href="#">
        Second default crumb
      </BreadcrumbItem>
      <BreadcrumbItem href="#" selected={false}>
        Selected crumb
      </BreadcrumbItem>
    </Breadcrumbs>
  );

  // Run tests on the fireEvents on BreadcrumbAnchor
  const breadcrumbAnchor = container.querySelector('a');

  fireEvent.click(breadcrumbAnchor);
  fireEvent.mouseDown(breadcrumbAnchor);

  // get rendered Breadcrumb elements to test against
  const firstDefaultBreadcrumb = getByText('First default crumb');
  const secondDefaultBreadcrumb = getByText('Second default crumb');
  const disabledBreadcrumb = getByText('Disabled crumb');
  const selectedBreadcrumb = getByText('Selected crumb');

  // Breadcrumb elements that are "disabled" and the last "selected" Breadcrumb should be <span> elements
  expect(disabledBreadcrumb.tagName.toLowerCase()).toBe('span');
  expect(disabledBreadcrumb).toHaveStyleRule('color', '#6f7171');

  const firstDefaultBreadcrumbStyless = window.getComputedStyle(
    firstDefaultBreadcrumb
  )._values;

  expect(firstDefaultBreadcrumbStyless['font-family']).toBe('Verizon-NHG-eTX');
  expect(firstDefaultBreadcrumbStyless['font-size']).toBe('0.75rem');
  expect(firstDefaultBreadcrumbStyless['font-weight']).toBe('400');
  expect(firstDefaultBreadcrumbStyless['line-height']).toBe('1rem');
  expect(firstDefaultBreadcrumb).toHaveStyleRule('color', '#000000');
  expect(selectedBreadcrumb.tagName.toLowerCase()).toBe('a');
  const selectedBreadcrumbStyles = window.getComputedStyle(selectedBreadcrumb)
    ._values;
  expect(selectedBreadcrumb).toHaveStyleRule('color', '#000000');
  expect(selectedBreadcrumbStyles['font-weight']).toBe('400');

  // Breadcrumb elements that have not been "disabled" and the last Breadcrumb should be <a> elements
  expect(firstDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
  expect(firstDefaultBreadcrumb).toHaveStyleRule('color', '#000000');

  expect(firstDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
  expect(secondDefaultBreadcrumb.tagName.toLowerCase()).toBe('a');
});
