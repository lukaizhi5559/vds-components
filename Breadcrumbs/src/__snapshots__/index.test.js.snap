// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs> 1`] = `
.c7 {
  font-size: 0.75rem;
  font-weight: 400;
  line-height: 1rem;
  font-family: Verizon-NHG-eTX,Helvetica,Arial,Sans-serif;
  color: #6f7171;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8 {
  font-size: 0.75rem;
  font-weight: 700;
  line-height: 1rem;
  font-family: Verizon-NHG-eTX,Helvetica,Arial,Sans-serif;
  color: #000000;
  margin: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6:active,
.c6:focus,
.c6:hover {
  outline: none;
}

.c3 {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: color 0.15s ease-out;
  transition: color 0.15s ease-out;
  position: relative;
  cursor: pointer;
  font-family: Verizon-NHG-eTX;
  font-size: 0.75rem;
  font-weight: 400;
  line-height: 1rem;
  color: #000000;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  box-sizing: border-box;
}

.c3:focus:not(:hover) {
  outline: none;
}

.c3:focus:not(:hover)::before {
  border: 0.0625rem dashed #000000;
  border-radius: 2px;
  content: '';
  height: calc(100% + 0.1875rem);
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  width: calc(100% + 0.25rem);
}

.c3:active {
  color: #6f7171;
}

.c3:hover:not(:active)::before {
  background: #000000;
  position: absolute;
  box-shadow: none;
  content: '';
  height: 1px;
  bottom: 0;
  width: 100%;
  z-index: 1;
}

.c4 {
  padding-right: 0.25rem;
  padding-left: 0.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  outline: none;
}

.c5 {
  color: #000000;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #6f7171;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 2px;
  margin-top: 2px;
}

.c0 {
  display: block;
  font-weight: 400;
  font-style: normal;
  color: #000000;
  line-height: 1.25;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 0;
  list-style-type: decimal;
  margin: 0;
}

<nav
  aria-label="Breadcrumb"
  class="c0"
>
  <ol
    class="c1"
  >
    <li
      class="c2"
    >
      <a
        aria-label="Go to Crumb 1"
        class="c3"
        color="#000000"
        href="#"
        role="link"
        tabindex="0"
        target="_self"
      >
        First default crumb
      </a>
      <div
        aria-hidden="true"
        class="c4"
        tabindex="-1"
      >
        <span
          aria-hidden="true"
          class="c5"
          color="#000000"
          tabindex="-1"
        >
          /
        </span>
      </div>
    </li>
    <li
      class="c2"
    >
      <span
        aria-hidden="false"
        class="c6 c7"
      >
        Disabled crumb
      </span>
      <div
        aria-hidden="true"
        class="c4"
        tabindex="-1"
      >
        <span
          aria-hidden="true"
          class="c5"
          color="#000000"
          tabindex="-1"
        >
          /
        </span>
      </div>
    </li>
    <li
      class="c2"
    >
      <a
        aria-label="Go to Crumb 3"
        class="c3"
        color="#000000"
        href="#"
        role="link"
        tabindex="0"
        target="_self"
      >
        Second default crumb
      </a>
      <div
        aria-hidden="true"
        class="c4"
        tabindex="-1"
      >
        <span
          aria-hidden="true"
          class="c5"
          color="#000000"
          tabindex="-1"
        >
          /
        </span>
      </div>
    </li>
    <li
      class="c2"
    >
      <span
        aria-hidden="false"
        class="c6 c8"
      >
        Selected crumb
      </span>
    </li>
  </ol>
</nav>
`;
